packages/doenetml-worker-javascript/src/utils/dast/sugar.ts

```diff
@@ -0,0 +1,334 @@
+import { deepClone } from "@doenet/utils";
+import { ComponentInfoObjects } from "../componentInfoObjects";
+import {
+    ErrorRecord,
+    isSerializedAttribute,
+    isSerializedComponent,
+    PrimitiveAttributeValue,
+    SerializedComponent,
+    WarningRecord,
+} from "./types";
+import { findPreSugarIndsAndMarkFromSugar } from "../expandDoenetML";
+import {
+    isUnflattenedAttribute,
+    isUnflattenedComponent,
+} from "./intermediateTypes";
+import {
+    expandAllUnflattenedAttributes,
+    expandUnflattenedToSerializedComponents,
+} from "./convertNormalizdDast";
+import { convertToSerializedErrorComponent } from "./errors";
+
+/**
+ * If a component class has specified sugar instructions,
+ * then apply those instructors to the children,
+ * potentially creating new children and attributes.
+ */
+export function applySugar({
+    serializedComponents,
+    parentParametersFromSugar = {},
+    parentAttributes = {},
+    componentInfoObjects,
+    isAttributeComponent = false,
+}: {
+    serializedComponents: (string | SerializedComponent)[];
+    parentParametersFromSugar?: Record<string, any>;
+    parentAttributes?: Record<string, any>;
+    componentInfoObjects: ComponentInfoObjects;
+    isAttributeComponent?: boolean;
+}): {
+    components: (string | SerializedComponent)[];
+    errors: ErrorRecord[];
+    warnings: WarningRecord[];
+} {
+    let errors: ErrorRecord[] = [];
+    let warnings: WarningRecord[] = [];
+
+    const newComponents: (string | SerializedComponent)[] = [];
+
+    for (let component of serializedComponents) {
+        if (typeof component === "string") {
+            newComponents.push(component);
+            continue;
+        }
+
+        try {
+            let componentType = component.componentType;
```

Comment:
I assume you've been in Rust mode, and most of these are meant to be `const`?

Reply:
This code was copied and pasted, and lightly edited from code I had written before I knew you were supposed to use `const`.